### PR TITLE
add weight cluster feature

### PR DESCRIPTION
**Pull request by endless001**
Add weighted clustering feature to yarp

### CODE REVIEW

**Comment by Tratcher** [comment]
Please don't submit PRs without first filing an issue to discuss the requirements and design. There are many variations of this scenario and many ways to implement it.

File: src/ReverseProxy/Model/ProxyPipelineInitializerMiddleware.cs [hunk 1]
```diff
@@ -39,6 +40,12 @@ public Task Invoke(HttpContext context)
         var route = endpoint.Metadata.GetMetadata<RouteModel>()
             ?? throw new InvalidOperationException($"Routing Endpoint is missing {typeof(RouteModel).FullName} metadata.");
 
+        var weightClusters = route.WeightClusters;
+
+        if (weightClusters.Any())
+        {
+            route.SetCluster(PickCluster(weightClusters));
```
Comment (Tratcher) [inline src/ReverseProxy/Model/ProxyPipelineInitializerMiddleware.cs:47]: The route object is shared across all requests, you can't change it per request.

File: src/ReverseProxy/Model/WeightedList.cs [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    }
+
+    public void Clear()
+    {
+        _list.Clear();
+        _weights.Clear();
+        Recalculate();
+    }
+
+    public T Next()
```
Comment (Tratcher) [inline src/ReverseProxy/Model/WeightedList.cs:46]: This would need to be thread safe.

**Comment by endless001** [comment]
Sorry, I submitted the "PR" without any discussion. I have seen that many reverse proxy tools have built-in weight clustering function, so I want to add this function to Yarp as well.

**Comment by karelz** [comment]
@endless001 as mentioned above, we should first discuss the feature and its design in an issue. Closing the PR for now. We can reopen it once we have feature agreement.

**Pull request closed without merging**